kvbc/test/sparse_merkle/internal_node_property_tests.cpp

```diff
@@ -0,0 +1,347 @@
+};
+
+}  // namespace rc
+
+// Return a vector of up to 16 unique nibbles.
+rc::Gen<std::vector<Nibble>> genUniqueNibbles() {
+  auto vec_nibble_gen = rc::gen::unique<std::vector<Nibble>>(rc::gen::arbitrary<Nibble>());
+  return rc::gen::resize(16, vec_nibble_gen);
+}
+
+// Return a generator for the depth of a BatchedInternalNode in a tree
+rc::Gen<size_t> genDepth() { return rc::gen::inRange<size_t>(0ul, Hash::MAX_NIBBLES); }
+
+rc::Gen<std::vector<LeafChild>> genLeafChildrenWithSameVersion(rc::Gen<std::vector<LeafChild>> childGen) {
+  return rc::gen::apply(
+      [](Version version, std::vector<LeafChild> children) {
+        std::vector<LeafChild> output;
+        for (const auto& child : children) {
+          output.emplace_back(LeafChild(child.hash, LeafKey(child.key.hash(), version)));
+        }
+        return output;
+      },
+      rc::gen::arbitrary<Version>(),
+      childGen);
+}
+
+// Create a generator to generate a maximum of 16 identical hashes, except that each one has a unique nibble at position
+// `depth`.
+rc::Gen<std::tuple<std::vector<Hash>, size_t>> genHashesWithUniqueNibblesAtGeneratedDepth() {
+  return rc::gen::apply(
+      [](const std::vector<Nibble>& nibbles, size_t depth) {
+        std::vector<Hash> hashes;
+        for (const auto& nibble : nibbles) {
+          auto digest = Hash::EMPTY_BUF;
+          setNibble(depth, digest, nibble);
+          hashes.push_back(Hash(digest));
+        }
+        return std::make_tuple(hashes, depth);
+      },
+      genUniqueNibbles(),
+      genDepth());
+}
+
+rc::Gen<std::tuple<std::vector<LeafChild>, size_t>> genKeysWithMatchingPrefixesUpToDepth(
+    rc::Gen<std::vector<LeafChild>> childGen) {
+  return rc::gen::apply(
+      [](auto children, size_t depth) {
+        for (auto& child : children) {
+          auto new_hash = child.key.hash();
+          for (auto i = 0ul; i < depth; i++) {
+            new_hash.setNibble(i, PLACEHOLDER_HASH.getNibble(i));
+            child.key = LeafKey(new_hash, child.key.version());
+          }
+        }
+        return std::make_tuple(children, depth);
+      },
+      childGen,
+      genDepth());
+}
+
+RC_GTEST_PROP(batched_internal_node_properties,
+              insert_on_empty_node_contains_leaf_child_at_depth_1,
+              (const LeafChild& child)) {
+  auto node = BatchedInternalNode();
+  auto result = node.insert(child, 0, child.key.version());
+  RC_ASSERT(std::holds_alternative<BatchedInternalNode::InsertComplete>(result));
+
+  // There is no stale leaf overwritte, since we insert a new value to an empty node
+  // RC_ASSERT_FALSE does compile with GCC :(
+  RC_ASSERT(std::get<BatchedInternalNode::InsertComplete>(result).stale_leaf.has_value() == false);
+
+  // Only the inserted child exists
+  RC_ASSERT(1u == node.numLeafChildren());
+
+  // The root is the only internal child
+  RC_ASSERT(1u == node.numInternalChildren());
+
+  // The LeafChild is directly under the root
```

Comment:
[Minor]: I think we can use `&&` in this particular test as they should both not be internal nodes (either a leaf or empty).

Reply:
Great catch! I wasn't thinking about empty nodes. Will fix.